<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
    "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
<meta charset="utf-8">
<title>Issue 3229: &sect;[res.on.exception.handling]#3 cannot apply to types with implicitly declared destructors</title>
<style type="text/css">
  p {text-align:justify}
  li {text-align:justify}
  blockquote.note
  {
    background-color:#E0E0E0;
    padding-left: 15px;
    padding-right: 15px;
    padding-top: 1px;
    padding-bottom: 1px;
  }
  ins {background-color:#A0FFA0}
  del {background-color:#FFA0A0}
  table {border-collapse: collapse;}
  @media (prefers-color-scheme: dark) {
     html {
        color: #ddd;
        background-color: black;
     }
     ins {
        background-color: #225522
     }
     del {
        background-color: #662222
     }
     a {
        color: #6af
     }
     a:visited {
        color: #6af
     }
     blockquote.note
     {
        background-color: rgba(255, 255, 255, .10)
     }
  }
</style>
</head>
<body>
<hr>
<h3><a name="3229" href="lwg-active.html#3229">3229.</a> &sect;[res.on.exception.handling]#3 cannot apply to types with implicitly declared destructors</h3>
<p><b>Section:</b> 16.4.6.13 <a href="https://timsong-cpp.github.io/cppwp/res.on.exception.handling">[res.on.exception.handling]</a> <b>Status:</b> <a href="lwg-active.html#New">New</a>
 <b>Submitter:</b> Nevin Liber <b>Opened:</b> 2019-06-28 <b>Last modified:</b> 2023-01-29 11:23:47 UTC</p>
<p><b>Priority: </b>3
</p>
<p><b>View other</b> <a href="lwg-index-open.html#res.on.exception.handling">active issues</a> in [res.on.exception.handling].</p>
<p><b>View all other</b> <a href="lwg-index.html#res.on.exception.handling">issues</a> in [res.on.exception.handling].</p>
<p><b>View all issues with</b> <a href="lwg-status.html#New">New</a> status.</p>
<p><b>Discussion:</b></p>
<p>
16.4.6.13 <a href="https://timsong-cpp.github.io/cppwp/res.on.exception.handling">[res.on.exception.handling]</a>#3 says:
</p>
<blockquote><p>
Destructor operations defined in the C++ standard library shall not throw exceptions.
Every destructor in the C++ standard library shall behave as if it had a non-throwing
exception specification.
</p></blockquote>
<p>
However, types like <tt>pair</tt> and <tt>array</tt> have implicitly declared destructors,
where 14.5 <a href="https://timsong-cpp.github.io/cppwp/except.spec">[except.spec]</a>#8 applies:
</p>
<blockquote><p>
The exception specification for an implicitly-declared destructor, or a destructor without
a <i>noexcept</i>-specifier, is potentially-throwing if and only if any of the destructors
for any of its potentially constructed subobjects is potentially-throwing or the destructor
is virtual and the destructor of any virtual base class is potentially throwing.
</p></blockquote>
<p>
We can relax 16.4.6.13 <a href="https://timsong-cpp.github.io/cppwp/res.on.exception.handling">[res.on.exception.handling]</a> to only cover explicitly declared
destructors because if they hold a user type where the destructor throws then we get UB
from 16.4.5.8 <a href="https://timsong-cpp.github.io/cppwp/res.on.functions">[res.on.functions]</a>#2:
</p>
<blockquote><p>
In particular, the effects are undefined in the following cases: [&hellip;]
</p>
<ol style="list-style-type:none">
<li><p>[&hellip;]</p></li>
<li><p>(2.4) &mdash; if any replacement function or handler function or destructor operation
exits via an exception, unless specifically allowed in the applicable
<i>Required behavior:</i> paragraph.</p></li>
<li><p>[&hellip;]</p></li>
</ol>
</blockquote>
<p>
and the referred to UB happens before [res.on.exception.handling] could apply.
</p>

<p><i>[2019-07 Issue Prioritization]</i></p>

<p>Priority to 3 after discussion on the reflector.</p>

<p><i>[2023-01-29; Daniel comments]</i></p>

<p>
This issue has very much overlap with LWG <a href="3854">3854</a>.
</p>


<p id="res-3229"><b>Proposed resolution:</b></p>
<p>This wording is relative to <a href="https://wg21.link/n4820">N4820</a>.</p>

<ol>
<li><p>Modify 16.4.6.13 <a href="https://timsong-cpp.github.io/cppwp/res.on.exception.handling">[res.on.exception.handling]</a> as indicated:</p>

<blockquote><p>
-3- Destructor operations defined in the C++ standard library shall not throw exceptions.
Every <ins>explicitly declared</ins> destructor in the C++ standard library shall behave
as if it had a non-throwing exception specification.
</p></blockquote>
</li>
</ol>




</body>
</html>
